File: matching.cpp
/**
 * \file matching.cpp
 *
 * \authors Mackenzie Kong-Sivert
 *
 * \brief Implementation of the model described in Akbarpour 2014
 */

#include "matching.hpp"

#include <math.h>
#include <stdlib.h>
#include <string>
#include <sstream>
#include <cstddef>
#include <string>
#include <vector>
#include <iostream>
#include <fstream>
#include <list>
#include <chrono>
#include <random>

Agent::Agent() :
    lifespan_{0},
    delta_{0},
    elapsed_time_{0},
    matched_{false}
{
    // Nothing to do here
}

Agent::Agent(uint64_t lifespan, float delta) :
    lifespan_{lifespan},
    delta_{delta},
    elapsed_time_{0},
    matched_{false}
{
    // Nothing left to do
}

bool Agent::time_step()
{
    ++elapsed_time_;
    return elapsed_time_==lifespan_;
}

float Agent::utility()
{
    return (matched_ ? exp(-1*delta_*elapsed_time_) : 0);
}

void Agent::match()
{
    matched_ = true;
}

Market::Market() :
    m_{0},
    d_{0},
    delta_{0},
    utility_total_{0}
{
    // Nothing left to do
}

Market::Market(uint64_t lambda, uint64_t m, uint64_t d, uint64_t delta, bool greedy) :
    m_{m},
    d_{d},
    delta_{delta},
    greedy_{greedy},
    utility_total_{0}
{
    lifespan_dist_ = std::poisson_distribution<int>(lambda);
    new_agent_dist_ = std::poisson_distribution<int>(m);
    uint64_t seed = std::chrono::system_clock::now().time_since_epoch().count();
    generator_ = std::default_random_engine(seed);
}

void Market::change_strategy(std::string strategy)
{
    if (strategy == "greedy")
    {
        greedy_ = true;
    }
    else if (strategy == "patient")
    {
        greedy_ = false;
    }
    else
    {
        greedy_ = !greedy_;
    }
}

float Market::utility()
{
    return utility_total_;
}

void Market::add_agent(Agent* a)
{
    std::list<Agent*> new_node {a};
    for (auto itr = compat_graph_.begin(); itr != compat_graph_.end(); ++itr)
    {
        if (rand()%m_ < d_) // probability of compatibility = d/m
        {
            new_node.push_back(itr->front()); // add connection
            itr->push_back(a);
        }
    }
    compat_graph_.push_back(new_node);
}

void Market::remove_agent(Agent* a)
{
    // find and remove the row for a
    auto a_itr = compat_graph_.end();
    for (auto itr = compat_graph_.begin(); itr != compat_graph_.end(); ++itr)
    {
        if (itr->front() == a)
        {
            a_itr = itr;
            break;
        }
    }
    if (a_itr != compat_graph_.end())
    {
        compat_graph_.erase(a_itr);
    }
    else
    {
        // throw an error or something
    }

    // find and remove a from all remaining rows
    for (auto itr1 = compat_graph_.begin(); itr1 != compat_graph_.end(); ++itr1)
    {
        auto a_itr = itr1->end();
        for (auto itr2 = itr1->begin(); itr2 != itr1->end(); ++itr2)
        {
            if (*itr2 == a)
            {
                a_itr = itr2;
                break;
            }
        }
        if (a_itr != itr1->end())
        {
            itr1->erase(a_itr);
        }
    }
}

bool Market::try_match(Agent* a)
{
    for (auto itr = compat_graph_.begin(); itr != compat_graph_.end(); ++itr)
    {
        if (*itr->begin() == a)
        {
            if (itr->size() > 1) // if there are any compatible agents, match one
            {
                auto b_itr = ++itr->begin();
                Agent* b = *b_itr;
                a->match();
                utility_total_ += a->utility();
                b->match();
                utility_total_ += b->utility();
                remove_agent(a);
                remove_agent(b);
                return true;
            }
            else
            {
                return false;
            }
        }
    }
    // if it gets to this point, the agent somehow is not in the market
    std::cout << "Error: Specified agent was not in the graph" << std::endl;
    return false;
}

void Market::time_step()
{
    std::list<Agent*> to_remove;
    for (auto itr = compat_graph_.begin(); itr != compat_graph_.end(); ++itr)
    {
        Agent* a = itr->front();
        if (a->time_step())
        {
            if (greedy_)
            {
                to_remove.push_back(a); // mark critical agents
            }
            else // under a patient strategy, try to match the agent at the end
            {
                bool matched = try_match(a);
                if (!matched)
                {
                    to_remove.push_back(a); // mark unmatched critical agents
                }
            }
        }
    }

    // erase all critical agents
    for (auto itr = to_remove.begin(); itr != to_remove.end(); ++itr)
    {
        remove_agent(*itr);
    }

    // get the number of agents to add from the poisson distribution
    uint64_t seed = std::chrono::system_clock::now().time_since_epoch().count();
    std::default_random_engine generator (seed);
    uint64_t new_agents = new_agent_dist_(generator_);
    for (uint64_t i = 0; i < new_agents; ++i)
    {
        uint64_t lifespan = lifespan_dist_(generator_);
        Agent* new_a = new Agent(lifespan, delta_);
        add_agent(new_a);
        if (greedy_)
        {
            try_match(new_a);
        }
    }
}

int main()
{
    // Constants
    uint64_t T = 20;
    uint64_t lambda = 4;
    uint64_t m = 10;
    uint64_t d = 5;
    uint64_t delta = 0;
    bool greedy = true;

    Market market(lambda, m, d, delta, greedy);
    for (uint64_t i = 0; i < T; ++i)
    {
        market.time_step();
        std::cout << "Total utility: " << market.utility() << std::endl;
    }
    return 0;
}

File: matching.py
import numpy as np
import math
import random
import sys
import matplotlib.pyplot as plt
from tqdm import tqdm

class Agent:
    def __init__(self, lifespan, delta):
        self._lifespan = lifespan
        self._delta = delta
        self._elapsed_time = 0
        self._matched = False

    def time_step(self):
        self._elapsed_time +=1
        return (self._elapsed_time == self._lifespan)

    def utility(self):
        return np.exp(-1*self._delta*self._elapsed_time) if self._matched else 0

    def match(self):
        self._matched = True

class Market:
    def __init__(self, lamb, m, d, delta, greedy) :
        self._m = m
        self._d = d
        self._delta = delta
        self._lambda = lamb
        self._greedy = greedy
        self._utility_total = 0
        self._unmatched = [0,0]
        self._compat_graph = []

    def change_strategy(self, strategy):
        if (strategy == "greedy"):
            self._greedy = true
        elif (strategy == "patient"):
            self._greedy = false
        else:
            self._greedy = not self._greedy

    def utility(self):
        return self._utility_total

    def loss(self):
        return self._unmatched[0] / self._unmatched[1] if self._unmatched[1] != 0 else 0

    def add_agent(self, a):
        new_node = [a]
        for node in self._compat_graph:
            #probability of compatibility = d/m
            if (random.randint(0,self._m) < self._d):
                new_node += [node[0]] # add connection
                node += [a]
        self._compat_graph+=[new_node]
        self._unmatched[0] += 1
        self._unmatched[1] += 1

    def remove_agent(self, a):
        # find and remove the row for a
        a_ind = sys.maxsize
        for i in range(len(self._compat_graph)):
            if (self._compat_graph[i][0] == a):
                a_ind = i
                break
        if a_ind != sys.maxsize:
            del self._compat_graph[a_ind]
        else:
            print("Error: Specified agent was not in the graph")
            return

        # find and remove a from all remaining rows
        for node in self._compat_graph: 
            if a in node:
                node.remove(a)

    def try_match(self, a):
        for node in self._compat_graph:
            if (node[0] == a):
                if (len(node) > 1): # if there are any compatible agents, match one
                    b = node[1]
                    a.match()
                    self._utility_total += a.utility()
                    b.match()
                    self._utility_total += b.utility()
                    self.remove_agent(a)
                    self.remove_agent(b)
                    self._unmatched[0] -= 2
                    return True
                else:
                    return False
        # if it gets to this point, the agent somehow is not in the market
        print("Error: Specified agent was not in the graph")
        return False

    def time_step(self):
        # get the number of agents to add from the poisson distribution
        new_agents = np.random.poisson(self._m, 1)[0]
        for i in range(new_agents):
            lifespan = np.random.poisson(self._lambda, 1)[0]
            new_a = Agent(lifespan, self._delta)
            self.add_agent(new_a)
            if (self._greedy):
                self.try_match(new_a)

        to_remove = []
        for node in self._compat_graph:
            a = node[0]
            if (a.time_step()):
                if (self._greedy):
                    to_remove+=[a] # mark critical agents
                else: # under a patient strategy, try to match the agent at the end
                    matched = self.try_match(a)
                    if (not matched):
                        to_remove+=[a] # mark unmatched critical agents

        # erase all critical agents
        for a in to_remove:
            self.remove_agent(a)

def main():
    T = 2000
    lamb = 4
    m = 10
    d = 5
    delta = 0
    greedy = True

    data_greedy = []
    for i in range(2,11):
        all_losses = []
        print('Running trial: d = ' + str(i))
        for j in tqdm(range(50)):
            market = Market(lamb, m, i, delta, True)
            for k in range(T):
                market.time_step()
            all_losses += [market.loss()]
        avg_loss = np.mean(all_losses)
        data_greedy += [(1/(2*i+1), avg_loss, math.log(2)/i)]    
        # (theoretical) lower bound, (theoretical) upper bound, average over 50 trials

    data_patient = []
    for i in range(2,11):
        all_losses = []
        print('Running trial: d = ' + str(i))
        for j in tqdm(range(50)):
            market = Market(lamb, m, i, delta, False)
            for k in range(T):
                market.time_step()
            all_losses += [market.loss()]
        avg_loss = np.mean(all_losses)
        data_patient += [(np.exp(-1*i)/(i+1), avg_loss, np.exp(-1*i/2)/2)]    
        # (theoretical) lower bound, (theoretical) upper bound, average over 50 trials

    #plot data
    for i in range(2):
        data = (data_greedy if i == 0 else data_patient)
        fig = plt.figure()
        ax1 = fig.add_subplot()
        ax1.set_ylabel('Loss')
        ax1.set_xlabel('d')
        strategy = '(Greedy)' if i == 0 else '(Patient)'
        ax1.set_title('Effect of d on Loss '+ strategy)
        x_values = range(2,11)
        lower = [x[0] for x in data]
        avg = [x[1] for x in data]
        upper = [x[2] for x in data]
        plt.plot(x_values, lower, label='Lower bound', color='darkgreen')
        plt.plot(x_values, avg, label='Average loss', color='navy')
        plt.plot(x_values, upper, label='Upper bound', color='firebrick')
        plt.legend()
        plt.show()


if __name__ == "__main__":
    main()